File: src/Annotation.cc
//
// Created by red on 9/11/17.
//

#include "Annotation.h"

Annotation::Annotation(const CallbackInfo& info)
  : ObjectWrap(info)
{
  if (info.Length() < 3) {
    throw Napi::Error::New(info.Env(),
                           "Annotation constructor requires three arguments of "
                           "type: Page, NPdfAnnotation, Rect");
  }
  auto pageObj = info[0].As<Object>();
  int npdfAnnotation = info[1].As<Number>();
  auto rectObj = info[2].As<Object>();
  auto page = Page::Unwrap(pageObj);
  auto rect = Rect::Unwrap(rectObj);
  doc = page->GetDocument();
  annot = page->GetPage()->CreateAnnotation(
    static_cast<PoDoFo::EPdfAnnotation>(npdfAnnotation), rect->GetRect());
}

void
Annotation::SetFlags(const CallbackInfo& info, const Napi::Value& value)
{
  if (value.IsNumber() == false) {
    throw Napi::TypeError::New(
      info.Env(), "SetFlag must be an instance of NpdfAnnotationType");
  }
  int jsValue = info[0].As<Number>();
  auto flag = static_cast<PoDoFo::EPdfAnnotationFlags>(jsValue);
  annot->SetFlags(flag);
}

Napi::Value
Annotation::GetFlags(const CallbackInfo& info)
{
  return Napi::Number::New(info.Env(), annot->GetFlags());
}

Napi::Value
Annotation::HasAppearanceStream(const CallbackInfo& info)
{
  return Napi::Boolean::New(info.Env(), annot->HasAppearanceStream());
}

void
Annotation::SetBorderStyle(const CallbackInfo& info)
{
  if (info.Length() != 3) {
    throw Napi::Error::New(info.Env(),
                           "Border requires three arguments of type number."
                           "Number - horizontal corner radius"
                           "Number - vertical corner radius"
                           "Number - width");
  }
  double horizontal = info[0].As<Number>();
  double vertical = info[1].As<Number>();
  double width = info[2].As<Number>();
  annot->SetBorderStyle(horizontal, vertical, width);
}

void
Annotation::SetTitle(const CallbackInfo& info, const Napi::Value& value)
{
  if (value.IsString() == false) {
    throw Napi::TypeError::New(
      info.Env(), "SetTitle requires a single argument of type string.");
  }
  try {
    string title = info[0].As<String>().Utf8Value();
    annot->SetTitle(PdfString(title));
  } catch (PdfError& err) {
    stringstream msg;
    msg << "PoDoFo PdfError: " << err.GetError() << endl;
    throw Napi::Error::New(info.Env(), msg.str());
  }
}

Napi::Value
Annotation::GetTitle(const CallbackInfo& info)
{
  return Napi::String::New(info.Env(), annot->GetTitle().GetString());
}

void
Annotation::SetContent(const CallbackInfo& info, const Napi::Value& value)
{
  if (value.IsEmpty()) {
    throw Napi::Error::New(info.Env(),
                           "SetContent requires string \"value\" argument.");
  }
  string content = value.As<String>().Utf8Value();
  annot->SetContents(content);
}

Napi::Value
Annotation::GetContent(const CallbackInfo& info)
{
  return Napi::String::New(info.Env(), annot->GetContents().GetString());
}

void
Annotation::SetDestination(const CallbackInfo& info, const Napi::Value& value)
{
  throw Napi::Error::New(info.Env(), "Not implemented.");
}

Napi::Value
Annotation::GetDestination(const CallbackInfo& info)
{
  throw Napi::Error::New(info.Env(), "Not implemented.");
}

Value
Annotation::HasDestination(const CallbackInfo& info)
{
  throw Napi::Error::New(info.Env(), "Not implemented.");
}

void
Annotation::SetAction(const CallbackInfo& info, const Napi::Value& value)
{
  Object actionObj = value.As<Object>();
  int type = actionObj.Get("type").As<Number>();
  string uri = actionObj.Get("uri").As<String>().Utf8Value();
  PoDoFo::EPdfAction flag = static_cast<PoDoFo::EPdfAction>(type);
  PdfAction action(flag, doc);
  action.SetURI(uri);
  annot->SetAction(action);
}

Napi::Value
Annotation::GetAction(const CallbackInfo& info)
{
  PdfAction* currentAction = annot->GetAction();
  if (currentAction->HasScript()) {
    return Napi::String::New(info.Env(),
                             currentAction->GetScript().GetString());
  } else if (currentAction->HasURI()) {
    return Napi::String::New(info.Env(), currentAction->GetURI().GetString());
  } else {
    throw Napi::Error::New(info.Env(), "Action not set");
  }
}

Napi::Value
Annotation::HasAction(const CallbackInfo& info)
{
  return Napi::Boolean::New(info.Env(), annot->HasAction());
}

void
Annotation::SetOpen(const CallbackInfo& info, const Napi::Value& value)
{
  if (value.IsBoolean() == false) {
    throw Napi::Error::New(info.Env(), "Requires Boolean type");
  }
  annot->SetOpen(value.As<Boolean>());
}

Napi::Value
Annotation::GetOpen(const CallbackInfo& info)
{
  return Napi::Boolean::New(info.Env(), annot->GetOpen());
}

void
Annotation::SetColor(const CallbackInfo& info, const Napi::Value& value)
{
  if (value.IsArray()) {
    Array jsValue = value.As<Array>();
    int rgb[3];
    for (uint8_t i = 0; i < jsValue.Length(); i++) {
      rgb[i] = jsValue.Get(i).As<Number>();
    }
    annot->SetColor(rgb[0], rgb[1], rgb[2]);
  } else {
    throw Napi::TypeError::New(info.Env(),
                               "Requires RGB color: [Number, Number, Number]");
  }
}

Napi::Value
Annotation::GetColor(const CallbackInfo& info)
{
  Array rgbArray = Napi::Array::New(info.Env());
  PdfArray pdfRgb = annot->GetColor();
  if (pdfRgb.size() != 3) {
  }
  double r = pdfRgb[0].GetNumber();
  double g = pdfRgb[1].GetNumber();
  double b = pdfRgb[2].GetNumber();
  rgbArray.Set(static_cast<uint8_t>(0), Napi::Number::New(info.Env(), r));
  rgbArray.Set(static_cast<uint8_t>(1), Napi::Number::New(info.Env(), g));
  rgbArray.Set(static_cast<uint8_t>(2), Napi::Number::New(info.Env(), b));
  return rgbArray;
}

Napi::Value
Annotation::GetType(const CallbackInfo& info)
{
  string jsType;
  switch (annot->GetType()) {
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_3D: {
      jsType = "3D";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Caret: {
      jsType = "Caret";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Circle: {
      jsType = "Circle";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_FileAttachement: {
      jsType = "FileAttachment";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_FreeText: {
      jsType = "FreeText";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Highlight: {
      jsType = "HighLight";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Ink: {
      jsType = "Ink";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Line: {
      jsType = "Line";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Link: {
      jsType = "Link";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Movie: {
      jsType = "Movie";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_PolyLine: {
      jsType = "PolyLine";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Polygon: {
      jsType = "Polygon";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Popup: {
      jsType = "Popup";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_PrinterMark: {
      jsType = "PrintMark";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_RichMedia: {
      jsType = "RichMedia";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Screen: {
      jsType = "Screen";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Sound: {
      jsType = "Sound";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Square: {
      jsType = "Square";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Squiggly: {
      jsType = "Squiggly";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Stamp: {
      jsType = "Stamp";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_StrikeOut: {
      jsType = "StikeOut";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Text: {
      jsType = "Text";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_TrapNet: {
      jsType = "TrapNet";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Unknown: {
      jsType = "Unknown";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Underline: {
      jsType = "Underline";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Watermark: {
      jsType = "Watermark";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_WebMedia: {
      jsType = "WebMedia";
      break;
    }
    case PoDoFo::EPdfAnnotation::ePdfAnnotation_Widget: {
      jsType = "Widget";
      break;
    }
  }
  return Napi::String::New(info.Env(), jsType);
}

void
Annotation::SetQuadPoints(const CallbackInfo& info, const Napi::Value& value)
{}

Napi::Value
Annotation::GetQuadPoints(const CallbackInfo& info)
{}

void
Annotation::SetFileAttachment(const CallbackInfo& info)
{}

Napi::Value
Annotation::HasFileAttachment(const CallbackInfo& info)
{
  return Napi::Boolean::New(info.Env(), annot->HasFileAttachement());
}

File: lib/annotation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var NPdfAnnotationType;
(function (NPdfAnnotationType) {
    NPdfAnnotationType["Text"] = "Text";
    NPdfAnnotationType["Link"] = "Link";
    NPdfAnnotationType["FreeText"] = "FreeText";
    NPdfAnnotationType["Line"] = "Line";
    NPdfAnnotationType["Square"] = "Square";
    NPdfAnnotationType["Circle"] = "Circle";
    NPdfAnnotationType["Polygon"] = "Polygon";
    NPdfAnnotationType["PolyLine"] = "PolyLine";
    NPdfAnnotationType["Highlight"] = "Highlight";
    NPdfAnnotationType["Underline"] = "Underline";
    NPdfAnnotationType["Squiggly"] = "Squiggly";
    NPdfAnnotationType["StrikeOut"] = "StrikeOut";
    NPdfAnnotationType["Stamp"] = "Stamp";
    NPdfAnnotationType["Caret"] = "Caret";
    NPdfAnnotationType["Ink"] = "Ink";
    NPdfAnnotationType["Popup"] = "Popup";
    NPdfAnnotationType["FileAttachment"] = "FileAttachment";
    NPdfAnnotationType["Sound"] = "Sound";
    NPdfAnnotationType["Movie"] = "Movie";
    NPdfAnnotationType["Widget"] = "Widget";
    NPdfAnnotationType["Screen"] = "Screen";
    NPdfAnnotationType["PrinterMark"] = "PrinterMark";
    NPdfAnnotationType["TrapNet"] = "TrapNet";
    NPdfAnnotationType["Watermark"] = "Watermark";
    NPdfAnnotationType["_3D"] = "3D";
    NPdfAnnotationType["RichMedia"] = "RichMedia";
    NPdfAnnotationType["WebMedia"] = "WebMedia";
})(NPdfAnnotationType = exports.NPdfAnnotationType || (exports.NPdfAnnotationType = {}));
var NPdfAnnotation;
(function (NPdfAnnotation) {
    NPdfAnnotation[NPdfAnnotation["Text"] = 0] = "Text";
    NPdfAnnotation[NPdfAnnotation["Link"] = 1] = "Link";
    NPdfAnnotation[NPdfAnnotation["FreeText"] = 2] = "FreeText";
    NPdfAnnotation[NPdfAnnotation["Line"] = 3] = "Line";
    NPdfAnnotation[NPdfAnnotation["Square"] = 4] = "Square";
    NPdfAnnotation[NPdfAnnotation["Circle"] = 5] = "Circle";
    NPdfAnnotation[NPdfAnnotation["Polygon"] = 6] = "Polygon";
    NPdfAnnotation[NPdfAnnotation["PolyLine"] = 7] = "PolyLine";
    NPdfAnnotation[NPdfAnnotation["Highlight"] = 8] = "Highlight";
    NPdfAnnotation[NPdfAnnotation["Underline"] = 9] = "Underline";
    NPdfAnnotation[NPdfAnnotation["Squiggly"] = 10] = "Squiggly";
    NPdfAnnotation[NPdfAnnotation["StrikeOut"] = 11] = "StrikeOut";
    NPdfAnnotation[NPdfAnnotation["Stamp"] = 12] = "Stamp";
    NPdfAnnotation[NPdfAnnotation["Caret"] = 13] = "Caret";
    NPdfAnnotation[NPdfAnnotation["Ink"] = 14] = "Ink";
    NPdfAnnotation[NPdfAnnotation["Popup"] = 15] = "Popup";
    NPdfAnnotation[NPdfAnnotation["FileAttachement"] = 16] = "FileAttachement";
    NPdfAnnotation[NPdfAnnotation["Sound"] = 17] = "Sound";
    NPdfAnnotation[NPdfAnnotation["Movie"] = 18] = "Movie";
    NPdfAnnotation[NPdfAnnotation["Widget"] = 19] = "Widget";
    NPdfAnnotation[NPdfAnnotation["Screen"] = 20] = "Screen";
    NPdfAnnotation[NPdfAnnotation["PrinterMark"] = 21] = "PrinterMark";
    NPdfAnnotation[NPdfAnnotation["TrapNet"] = 22] = "TrapNet";
    NPdfAnnotation[NPdfAnnotation["Watermark"] = 23] = "Watermark";
    NPdfAnnotation[NPdfAnnotation["_3D"] = 24] = "_3D";
    NPdfAnnotation[NPdfAnnotation["RichMedia"] = 25] = "RichMedia";
    NPdfAnnotation[NPdfAnnotation["WebMedia"] = 26] = "WebMedia";
})(NPdfAnnotation = exports.NPdfAnnotation || (exports.NPdfAnnotation = {}));
var NpdfAnnotationFlag;
(function (NpdfAnnotationFlag) {
    NpdfAnnotationFlag[NpdfAnnotationFlag["Invisible"] = 1] = "Invisible";
    NpdfAnnotationFlag[NpdfAnnotationFlag["Hidden"] = 2] = "Hidden";
    NpdfAnnotationFlag[NpdfAnnotationFlag["Print"] = 4] = "Print";
    NpdfAnnotationFlag[NpdfAnnotationFlag["NoZoom"] = 8] = "NoZoom";
    NpdfAnnotationFlag[NpdfAnnotationFlag["NoRotate"] = 16] = "NoRotate";
    NpdfAnnotationFlag[NpdfAnnotationFlag["NoView"] = 32] = "NoView";
    NpdfAnnotationFlag[NpdfAnnotationFlag["ReadOnly"] = 64] = "ReadOnly";
    NpdfAnnotationFlag[NpdfAnnotationFlag["Locked"] = 128] = "Locked";
    NpdfAnnotationFlag[NpdfAnnotationFlag["ToggleNoView"] = 256] = "ToggleNoView";
    NpdfAnnotationFlag[NpdfAnnotationFlag["LockedContents"] = 512] = "LockedContents";
})(NpdfAnnotationFlag = exports.NpdfAnnotationFlag || (exports.NpdfAnnotationFlag = {}));
var NpdfAction;
(function (NpdfAction) {
    NpdfAction[NpdfAction["GoTo"] = 0] = "GoTo";
    NpdfAction[NpdfAction["GoToR"] = 1] = "GoToR";
    NpdfAction[NpdfAction["GoToE"] = 2] = "GoToE";
    NpdfAction[NpdfAction["Launch"] = 3] = "Launch";
    NpdfAction[NpdfAction["Thread"] = 4] = "Thread";
    NpdfAction[NpdfAction["URI"] = 5] = "URI";
    NpdfAction[NpdfAction["Sound"] = 6] = "Sound";
    NpdfAction[NpdfAction["Movie"] = 7] = "Movie";
    NpdfAction[NpdfAction["Hide"] = 8] = "Hide";
    NpdfAction[NpdfAction["Named"] = 9] = "Named";
    NpdfAction[NpdfAction["SubmitForm"] = 10] = "SubmitForm";
    NpdfAction[NpdfAction["ResetForm"] = 11] = "ResetForm";
    NpdfAction[NpdfAction["ImportData"] = 12] = "ImportData";
    NpdfAction[NpdfAction["JavaScript"] = 13] = "JavaScript";
    NpdfAction[NpdfAction["SetOCGState"] = 14] = "SetOCGState";
    NpdfAction[NpdfAction["Rendition"] = 15] = "Rendition";
    NpdfAction[NpdfAction["Trans"] = 16] = "Trans";
    NpdfAction[NpdfAction["GoTo3DView"] = 17] = "GoTo3DView";
    NpdfAction[NpdfAction["RichMediaExecute"] = 18] = "RichMediaExecute";
})(NpdfAction = exports.NpdfAction || (exports.NpdfAction = {}));
var Annotation = /** @class */ (function () {
    function Annotation(_instance) {
        this._instance = _instance;
    }
    Object.defineProperty(Annotation.prototype, "quadPoints", {
        get: function () {
            throw Error("unimplemented");
        },
        set: function (value) {
            throw Error("unimplemented");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Annotation.prototype, "title", {
        get: function () {
            return this._instance.title;
        },
        set: function (value) {
            this._instance.title = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Annotation.prototype, "flag", {
        get: function () {
            return this._instance.flag;
        },
        set: function (value) {
            this._instance.flag = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Annotation.prototype, "color", {
        get: function () {
            return this._instance.color;
        },
        set: function (value) {
            var rgbErr = Error("RGB value must be an integer >= 0 || <= 256");
            if (value.length !== 3) {
                throw rgbErr;
            }
            value.forEach(function (e) {
                if (Number.isInteger(e) === false) {
                    throw rgbErr;
                }
                if (Number.isNaN(e)) {
                    throw rgbErr;
                }
            });
            this._instance.color = value;
        },
        enumerable: true,
        configurable: true
    });
    Annotation.prototype.hasAppearanceStream = function () {
        return this._instance.hasAppearanceStream();
    };
    Annotation.prototype.setBorderStyle = function (horizontalRadius, verticalRadius, width) {
        return this._instance.setBorderStyle(horizontalRadius, verticalRadius, width);
    };
    Annotation.prototype.hasDestination = function () {
        return this._instance.hasDestination();
    };
    Annotation.prototype.hasAction = function () {
        return this._instance.hasAction();
    };
    Annotation.prototype.getType = function () {
        return this._instance.getType();
    };
    Annotation.prototype.setFileAttachment = function () {
        throw new Error("Method not implemented.");
    };
    Annotation.prototype.hasFileAttachment = function () {
        return this._instance.hasFileAttachment();
    };
    return Annotation;
}());
exports.Annotation = Annotation;
